From c13ced2b011338f50340f06c8dd11d273c8fca5b Mon Sep 17 00:00:00 2001
From: Ilya Ledvich <ilya@compulab.co.il>
Date: Thu, 6 Oct 2016 15:02:09 +0300
Subject: [PATCH 15/57] HACK: usb: ehci-mx6: do not initialize otg2 port for
 iMX7 Solo

Number of ports that USB sub-system initializes is determined by
CONFIG_USB_MAX_CONTROLLER_COUNT that is statically defined in the
board config file. So we need some mechanism to determine number of
USB OTG ports in running time, because the iMX7 Dual SoC has two
OTG ports, while the iMX7 Solo has only one.

Signed-off-by: Ilya Ledvich <ilya@compulab.co.il>
---
 drivers/usb/host/ehci-mx6.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/usb/host/ehci-mx6.c b/drivers/usb/host/ehci-mx6.c
index f348ec9bca..7bb3e5cf73 100644
--- a/drivers/usb/host/ehci-mx6.c
+++ b/drivers/usb/host/ehci-mx6.c
@@ -338,6 +338,8 @@ int ehci_mx6_common_init(struct usb_ehci *ehci, int index)
 }
 
 #ifndef CONFIG_DM_USB
+#include <asm/arch/sys_proto.h>
+
 int ehci_hcd_init(int index, enum usb_init_type init,
 		struct ehci_hccr **hccr, struct ehci_hcor **hcor)
 {
@@ -346,6 +348,7 @@ int ehci_hcd_init(int index, enum usb_init_type init,
 	u32 controller_spacing = 0x200;
 #elif defined(CONFIG_MX7)
 	u32 controller_spacing = 0x10000;
+	u32 cpurev = get_cpu_rev();
 #endif
 	struct usb_ehci *ehci = (struct usb_ehci *)(USB_BASE_ADDR +
 		(controller_spacing * index));
@@ -354,6 +357,12 @@ int ehci_hcd_init(int index, enum usb_init_type init,
 	if (index > 3)
 		return -EINVAL;
 
+#if defined (CONFIG_MX7)
+	/* Initialize only one port for i.MX7Solo */
+	if (((cpurev & 0xFF000) >> 12) == MXC_CPU_MX7S)
+		if (index > 0)
+			return -ENODEV;
+#endif
 	ret = ehci_mx6_common_init(ehci, index);
 	if (ret)
 		return ret;
-- 
2.17.1

